@here/harp-mapview/lib/Utils.ts

```diff
@@ -242,53 +243,182 @@ export namespace MapViewUtils {
+        }
+
+        const ninetyRad = THREE.MathUtils.degToRad(90);
+        // The following terminology will be used:
+        // R = rotation target, i.e. the point about which we are rotating: `rotationTargetWorld`
+        // C = camera position
+        // M = map target, i.e. the point which the camera is looking at: `mapTargetWorld`
+        // The following variable `angleBetweenRotationTargetAndMapTarget` is the angle between the
+        // two rays C->R and C->M, i.e. the angle RCM. This angle remains constant when tilting with
+        // a fixed `offsetX` and `offsetY`. It is calculated by using the intersection of the two
+        // ways with the earth. Where:
+        // MRC is 90 + `angleBetweenNormals` - tilt at R.
+        // CMR is 90 + `tilt`
+        // RCM is 180 - CRM - RMC, aka `angleBetweenRotationTargetAndMapTarget`
+        // Note the use of `angleBetweenNormals` to ensure this works for spherical projections.
+        const angleBetweenRotationTargetAndMapTarget =
+            ninetyRad * 2 -
+            (ninetyRad + angleBetweenNormals - rotationCenterTilt + ninetyRad + tilt);
+
+        // This uses the same formula as above, just we change `tilt` to the maxTiltAngle,
+        // i.e we want to find the greatest angle at the rotation target that gives us the max
+        // angle at the map center target.
+        // i.e. CMR is 90 + `maxTiltAngle`
+        const maxRotationTargetAngle =
+            ninetyRad * 2 - angleBetweenRotationTargetAndMapTarget - ninetyRad - maxTiltAngle;
+
+        // Converting the `maxRotationTargetAngle` back to a tilt is as easy as subtracting it from
+        // 90 and the `angleBetweenNormals`, i.e. this gives us the maximum allowed tilt at R
+        // that satisfies the `maxTiltAngle` constraint.
+        const maxTilt = ninetyRad + angleBetweenNormals - maxRotationTargetAngle;
+
+        const clampedDeltaTiltMapTarget =
+            MathUtils.clamp(deltaTilt + rotationCenterTilt, 0, maxTilt) - rotationCenterTilt;
+
+        const maxRotationTiltAngle = THREE.MathUtils.degToRad(89);
+        // Here we clamp to 89 degrees, to prevent it intersecting with the world at 90. This is
+        // possible for example when the R position is near the horizon. If the angle RCM is say 5
+        // degrees, then an angle of say 89 degrees at R, plus 5 degrees means the tilt at M would
+        // be 84 degrees. If the `maxTiltAngle` is however 89, then we still have what seems like 5
+        // degrees at M, so, without this check, we would go to 90 degrees.
+        const clampedDeltaTiltRotationTarget =
+            MathUtils.clamp(deltaTilt + rotationCenterTilt, 0, maxRotationTiltAngle) -
+            rotationCenterTilt;
+
+        // Math.min doesn't work, because we need to keep the sign of the delta as well.
+        const clampedDeltaTilt = Math.min(
+            clampedDeltaTiltMapTarget,
+            clampedDeltaTiltRotationTarget
+        );
```

Comment:
Here you're doing:

clampedDeltaTilt = min(clampedDeltaTiltM,clampedDeltaTiltR) = 
min( clamp(deltaTilt + rotCenterTilt, 0, maxTilt) - rotCenterTilt, clamp(deltaTilt + rotCenterTilt, 0, maxRotTiltAngle) - rotationCenterTilt ) =
min( clamp(deltaTilt + rotCenterTilt, 0, maxTilt) , clamp(deltaTilt + rotCenterTilt, 0, maxRotTiltAngle) ) - rotCenterTilt

So this can be simplified to:

min( clamp(deltaTilt + rotCenterTilt, 0, min(maxTilt,maxRotTiltAngle)) - rotCenterTilt

Is this this intended? Does it make sense to compute clampedDeltaTiltMapTarget using rotationCenterTilt ?

Reply:
Nice, I hadn't considered to simplify this, but it is a good idea.

> Does it make sense to compute clampedDeltaTiltMapTarget using rotationCenterTilt ?

Yes, the `clampedDeltaTiltMapTarget` is named a bit incorrectly, it is the delta tilt clamped that, when applied to the rotation center, would give a correctly clamped value at `mapTargetWorld`